metrics/repl.lag/lag.go

```diff
@@ -16,32 +17,65 @@ import (
 const (
 	DOMAIN = "repl.lag"
 
-	OPT_SOURCE_ID            = "source-id"
-	OPT_SOURCE_ROLE          = "source-role"
-	OPT_TABLE                = "table"
-	OPT_WRITER               = "writer"
-	OPT_REPL_CHECK           = "repl-check"
-	OPT_REPORT_NO_HEARTBEAT  = "report-no-heartbeat"
-	OPT_REPORT_NOT_A_REPLICA = "report-not-a-replica"
-	OPT_NETWORK_LATENCY      = "network-latency"
+	OPT_HEARTBEAT_SOURCE_ID                = "source-id"
+	OPT_HEARTBEAT_SOURCE_ROLE              = "source-role"
+	OPT_HEARTBEAT_TABLE                    = "table"
+	OPT_WRITER                             = "writer"
+	OPT_REPL_CHECK                         = "repl-check"
+	OPT_REPORT_NO_HEARTBEAT                = "report-no-heartbeat"
+	OPT_REPORT_NOT_A_REPLICA               = "report-not-a-replica"
+	OPT_RENAME_DEFAULT_REPLICATION_CHANNEL = "rename-default-replication-channel"
+	OPT_NETWORK_LATENCY                    = "network-latency"
+
+	LAG_WRITER_BLIP = "blip"
+	LAG_WRITER_PFS  = "pfs"
+
+	// MySQL8LagQuery is the query to calculate approximate lag
+	// from replication worker stats in performance schema
+	// This is only available in MySQL 8 (and above) and when performance schema is enabled
+	MySQL8LagQuery = `WITH applier_latency AS (
+ SELECT TIMESTAMPDIFF(MICROSECOND, LAST_APPLIED_TRANSACTION_ORIGINAL_COMMIT_TIMESTAMP, LAST_APPLIED_TRANSACTION_END_APPLY_TIMESTAMP)/1000 as applier_latency_ms
+ FROM performance_schema.replication_applier_status_by_worker ORDER BY LAST_APPLIED_TRANSACTION_END_APPLY_TIMESTAMP DESC LIMIT 1
+), queue_latency AS (
+ SELECT MIN(
+ CASE
+  WHEN
+   LAST_QUEUED_TRANSACTION = 'ANONYMOUS' OR
+   LAST_APPLIED_TRANSACTION = 'ANONYMOUS' OR
+   GTID_SUBTRACT(LAST_QUEUED_TRANSACTION, LAST_APPLIED_TRANSACTION) = ''
+  THEN 0
+   ELSE
+   TIMESTAMPDIFF(MICROSECOND, LAST_APPLIED_TRANSACTION_IMMEDIATE_COMMIT_TIMESTAMP, NOW(3))/1000
+ END
+) AS queue_latency_ms,
+IF(MIN(TIMESTAMPDIFF(MINUTE, LAST_QUEUED_TRANSACTION_ORIGINAL_COMMIT_TIMESTAMP, NOW()))>1,'IDLE','ACTIVE') as queue_status
+FROM performance_schema.replication_applier_status_by_worker w
+JOIN performance_schema.replication_connection_status s ON s.channel_name = w.channel_name
+)
+SELECT IF(queue_status='IDLE',0,GREATEST(applier_latency_ms, queue_latency_ms)) as lagMs FROM applier_latency, queue_latency;`
+
+	defaultChannelName = "default"
```

Comment:
Empty string because Blip tries to mirror MySQL directly. It only changes stuff when MySQL is so odd that industry standard never uses the odd MySQL thing. But in this case, channel="" works because empty string is a valid and unique key in a `map[string]T`.